File: client/python/mesh/macro/spi.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#

from typing import Type, Any, Generic

from mesh.macro.ark import ark, T


class SPI(Generic[T]):

    def __init__(self, fn: T = None, name='', pattern='', priority=0, prototype=False):
        """
        Metadata annotation for Serial Peripheral Interface. Can be used with {@link ServiceLoader#load(Class)}
        or dependency injection at compile time and runtime time.
        """
        self.name = name
        self.pattern = pattern
        self.priority = priority
        self.prototype = prototype
        if name != '':
            self.name = name
            return
        if fn is None:
            return
        if type(fn) is str:
            self.name = fn
            return
        self.name = str(type(fn))
        self.kind = fn

    def __call__(self, *args, **kwargs) -> T:
        if not hasattr(self, 'kind'):
            self.kind = args[0]
        self.kind.__spi__ = self
        ark.register(spi, self.name, self.kind, self)
        return self.kind

    @staticmethod
    def get_macro(kind: Type[Any]) -> "SPI":
        if hasattr(kind, '__spi__'):
            return kind.__spi__
        return SPI()


def spi(fn: T = None, *, name='', pattern='', priority=0, prototype=False):
    return SPI(fn, name, pattern, priority, prototype)

File: client/java/src/main/java/com/be/mesh/client/annotate/SPI.java
/*
 * Copyright (c) 2000, 2023, ducesoft and/or its affiliates. All rights reserved.
 * DUCESOFT PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.annotate;

import java.lang.annotation.Documented;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.util.ServiceLoader;

import static java.lang.annotation.ElementType.*;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Metadata annotation for Serial Peripheral Interface. Can be used with {@link ServiceLoader#load(Class)} or
 * dependency injection at compile time and runtime time.
 * <p/>
 * Annotated on class or method will provider service as SPI.
 * <p/>
 * Example:
 * <pre>
 * &#064;SPI(value="name")
 * public class SPIService {
 *
 *     &#064;SPI(value="name")
 *     public SPIService spi(){...}
 * }
 * </pre>
 * Annotated on field or parameter will reference SPI service as reference for dependency injection.
 * <p/>
 * Example:
 * <pre>
 * public class Demo {
 *      &#064;SPI(value="name")
 *      private SPIService spi;
 *
 *      public void call(&#064;SPI(value="name") SPIService spi) {...}
 * }
 * </pre>
 *
 * @author coyzeng@gmail.com
 */
@Inherited
@Documented
@Retention(RUNTIME)
@Target({TYPE, METHOD, FIELD, PARAMETER})
public @interface SPI {

    /**
     * xxx.xxx.xxx
     */
    String value() default "";

    /**
     * Pattern match
     */
    String pattern() default "";

    /**
     * spi priority order by asc, highest is {@link Integer#MIN_VALUE}
     */
    int priority() default 0;

    /**
     * metadata for the spi.
     */
    boolean prototype() default false;

    /**
     * SPI alias.
     */
    String[] alias() default {};

    /**
     * metadata for the spi.
     */
    Class<?>[] meta() default {};

    /**
     * Exclude metadata.
     */
    String[] exclude() default {};

    /**
     * Enable the spi provider default.
     */
    boolean enable() default true;
}
